BUG: np.load does not handle empty array with an empty descr

liushapku opened the pull request:
The bug occurs since numpy 1.16. Before that empty descr corresponds to
`np.dtype([])`. This fixes the problem by following numpy 1.15's
behavior. #15396 

<!-- Please be sure you are following the instructions in the dev guidelines
http://www.numpy.org/devdocs/dev/development_workflow.html
-->

<!-- We'd appreciate it if your commit message is properly formatted
http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message
-->

liushapku committed:
BUG: np.load does not handle empty array with an empty descr

The bug occurs since numpy 1.16. Before that empty descr corresponds to
`np.dtype([])`. This fixes the problem by following numpy 1.15's
behavior.

eric-wieser commented on numpy/lib/format.py:
```diff
@@ -313,6 +313,9 @@ def descr_to_dtype(descr):
         offset += dt.itemsize
 
+    if not fields:
+        return numpy.dtype([])
```
This doesn't respect the itemsize.

I'd recommend fixing this by replacing `fields.append((name, dt, offset))` with `names.append(name)`, `formats.append(dt)`, etc

liushapku replied:
I tried the following
```
In [7]: np.dtype({'names': [], 'formats': [], 'titles': [], 'offsets': [], 'itemsize': 0})                                                                                                            
Out[7]: dtype([])

In [8]: np.dtype({'names': [], 'formats': [], 'titles': [], 'offsets': [], 'itemsize': 0})  == np.dtype([])                                                                                                   
Out[8]: True
```
So there is no difference.

eric-wieser replied:
`itemsize` might not be 0.

liushapku replied:
What exactly is the itemsize for an empty array with empty descr?

eric-wieser replied:
It depends on which empty descriptor. Your example has size 0, but you can construct one with any size you like using the dictionary form.

This code contains a loop that does just that.

eric-wieser commented on numpy/lib/format.py:
```diff
@@ -313,6 +313,9 @@ def descr_to_dtype(descr):
 
```
I'd suggest replacing the line above this with:
```python
title, name = name if isinstance(name, tuple) else (None, name)
names.append(name)
formats.append(dt)
titles.append(title)
offsets.append(offset)
```
and then removing the four lines before the return in this function. You'll want to replace `fields = []` with something like
`names, formats, titles, offsets = [], [], [], []`.

liushapku replied:
I actually tried exactly the same code. Then I found that is returns exactly same thing as `np.dtype([])`. So I came here and showed you the result of [this snippet](https://github.com/numpy/numpy/pull/15397#discussion_r370006812).

I will push the new code. You can decide which one to use

eric-wieser replied:
The test you're missing is `descr_to_dtype([('', 'V8')])`, which produced the wrong result with your original code.

liushapku committed:
BUG: np.load does not handle empty array with an empty descr

The bug occurs since numpy 1.16. Before that empty descr corresponds to
`np.dtype([])`. This fixes the problem by following numpy 1.15's
behavior. Code are slightly change to replace the use of zip

eric-wieser commented:
Can you add a new tests to this file?

https://github.com/numpy/numpy/pull/12358/files#diff-8cb06279635f588dc7f134a3a4d7d793
```python
dt6 = np.dtype({'names': [], 'formats': [], 'itemsize': 8})
```
would do the job

liushapku committed:
BUG: np.load does not handle empty array with an empty descr

add test for the bug fix

seberg commented:
Thanks all, was mostly waiting for the tests :).
